feat: add dart frog test, a different approach (#899)
VeryGoodOpenSource/dart_frog: packages/dart_frog_test/README.md

# Dart Frog Test

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![Powered by Mason](https://img.shields.io/endpoint?url=https%3A%2F%2Ftinyurl.com%2Fmason-badge)](https://github.com/felangel/mason)
[![License: MIT][license_badge]][license_link]

A testing library that makes it easy to test Dart Frog services. It offers helpers to mock requests as well as

> custom matchers in order to write easier to read expectations.

Reviewer: **Nit:**
```suggestion
custom matchers in order to write readable expectations.
```

## Installation 💻

**❗ In order to start using Dart Frog Test you must have the [Dart SDK][dart_install_link] installed on your machine.**

Add `dart_frog_test` to your `pubspec.yaml`:

```yaml
dependencies:
  dart_frog_test:
```

Install it:

```sh
dart pub get
```

## DartFrogTestContext

This class makes it simple to mock a `RequestContext` for a Dart Frog request handler. To use it, simply import it
and use its constructor and methods to create the mocker context.

A simple example:

```dart
// Mocking a get request, which is the default
import '../../../routes/users/[id].dart' as route;

test('returns ok', () {

>   final context - DartFrogTestContext(

Reviewer: ```suggestion
  final context = DartFrogTestContext(
```

    path: '/users/1',
  );

  final response = route.onRequest(context);
  expect(response.statusCode, equals(200));
});
```

If the route handler function reads a [dependency injected via context](https://dartfrog.vgv.dev/docs/basics/dependency-injection), that can also be mocked:

```dart
// Mocking a get request, which is the default
import '../../../routes/users/index.dart' as route;

test('returns ok', () {

>   final context - DartFrogTestContext(

Reviewer: ```suggestion
  final context = DartFrogTestContext(
```

    path: '/users',
  );

>   final userRepository = /** Create Mock */;

Reviewer: ```suggestion
  final userRepository = /* Create Mock */;
```

Author: I would rather not put mocktail here to not influence the reader. I would prefer people to use Mocktail, but Mockito would work too, so better not to influence IMO

  context.provide<UserRepository>(userRepository);

  final response = route.onRequest(context);
  expect(response.statusCode, equals(200));
});
```

> Check the `DartFrogTestContext` constructor for all the available context attributes that can be mocked.

Reviewer: What do you think of linking this to the [API reference hosted by pub.dev](https://pub.dev/documentation/dart_frog_test/latest/)?

Author: ```suggestion
Check the `DartFrogTestContext` [constructor](https://pub.dev/documentation/dart_frog_test/latest/) for all the available context attributes that can be mocked.
```

## Matchers

> This package also provide test matchers that can be used to do expectation/assertions on top of

Reviewer: ```suggestion
This package also provide test matchers that can be used to do expectation or assertions on top of
```

# Dart Frog Test

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![Powered by Mason](https://img.shields.io/endpoint?url=https%3A%2F%2Ftinyurl.com%2Fmason-badge)](https://github.com/felangel/mason)
[![License: MIT][license_badge]][license_link]

A testing library that makes it easy to test Dart Frog services. It offers helpers to mock requests as well as
custom matchers in order to write readable expectations.

> [!NOTE]
> This package is still experimental and although it is ready to be used, some/or all of its API

> > might change (with deprecations) in future versions.

Reviewer: Does the `[!NOTE]` render similarly in Pub as in GitHub?

Author: Not really sure, I know that github markdown is way advanced.

Reviewer: Let's find out then :eyes:
